### PR TITLE
[OBSOLETE] Add sandbox support for nsjail

### PR DESCRIPTION
**Pull request by apmorton**
You must compile [nsjail](https://github.com/google/nsjail) and manually put the resulting binary somewhere on your path - nsjail does not have a `make install` target.

Due to lack of seccomp policies I would not consider this ready to be deployed publicly - but it may be useful to have this temporarily deployed to beta to facilitate testing with all supported compilers.

# Notes
You must run the following as root after each system reboot to create the required cgroups before starting CE
```
cgcreate -a ubuntu:ubuntu -g memory,pids,cpu,net_cls:ce-sandbox
```

# TODO
- [ ] seccomp policies
- [ ] test with various languages to see if any runtimes get upset about lack of `/etc` and other directory structures
- [ ] execution support to allow compilers under nsjail
- [ ] figure out how wine will work under nsjail

# Test bench to explore sandbox environment
https://godbolt.org/z/a8eItT

Results running under nsjail locally
https://gist.github.com/apmorton/088daf802689fc55fcef0283d60cd8bc

### CODE REVIEW

**Comment by RabsRincon** [comment]
I think this would help closing #1412 and #1428 too, right?

**Comment by partouf** [comment]
> I think this would help closing #1412 and #1428 too, right?

The question is; what will the difference be between the issues closed and newly created because of this change ;)

**Comment by apmorton** [comment]
It definitely closes #1428, and would probably fix #1412 but I haven't been able to test that.

partouf is correct that switching up the jailing solution will probably open new bugs, but hopefully we can catch at least the most egregious of those on the beta instance.

Merging this shouldn't break anything, since the defaults have not been changed.

I can look at a PR to get nsjail compiled into the image so we can actually test this with all the compilers, but I will definitely need someones help actually testing things out since I don't have things configured on my end to run the image building process.

**Comment by mattgodbolt** [comment]
Thanks so much! this is great! Will take a proper look tomorrow, but thank you for doing this. Of course changing the solution we're using is a risk; but we've hit some real problems with `firejail` and this appears to be a better overall solution. We'll see how it goes! :)

File: lib/exec.js [hunk 1]
```diff
@@ -139,6 +139,41 @@ function executeDirect(command, args, options, filenameTransform) {
...
+
+function sandboxNsjail(command, args, options) {
+    logger.info("Sandbox execution via nsjail", {command, args});
+    const execPath = path.dirname(command);
+    const execName = path.basename(command);
+
+    const jailingOptions = withNsjailTimeout([
+        '--config', 'etc/nsjail/sandbox.cfg',
+        '--cwd', '/app',
+        '--bindmount', `${execPath}:/app`,
```
Comment (mattgodbolt) [inline lib/exec.js]: Cunning; this is how we map in the sandbox. Nice that there's already a `--cwd` :)

**Comment by mattgodbolt** [comment]
Forgive me: I didn't spot one of your TODOs was to get execution of compilers in too!

**Comment by mattgodbolt** [comment]
For what it's worth I'd happily merge this now to be a work in progress, but I don't want to lose your TODO list above. Thoughts? :)

**Comment by mattgodbolt** [comment]
Just tried this and it Just Worked™ - nice work @apmorton ! Will have a think about the TODOs you've got there (and I may well merge and create issues to track them for now)

File: etc/nsjail/sandbox.cfg [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+# must run following as root during system startup
+# cgcreate -a ubuntu:ubuntu -g memory,pids,cpu,net_cls:ce-sandbox
+cgroup_mem_parent: "ce-sandbox"
+cgroup_pids_parent: "ce-sandbox"
+cgroup_net_cls_parent: "ce-sandbox"
+cgroup_cpu_parent: "ce-sandbox"
+
+cgroup_mem_max: 209715200 # 200 MiB
+cgroup_pids_max: 4
```
[the remark this answers is not in the document]
Comment (apmorton) [reply etc/nsjail/sandbox.cfg]: each sandbox will get 4 PIDs.

Each sandbox will create a new cgroup as a child of the ce-sandbox cgroup and set it's own limits (200MB ram, 4 PIDs {which includes threads afaik}).

You could also set system wide limits for all sandboxes combine by putting limits on the ce-sandbox cgroup.

I planned to have a separate parent cgroup for compiler execution as well (ce-execution or something), so we could set different resource limits on each group.

**Comment by mattgodbolt** [comment]
We should chat about this at CppCon @apmorton  :)

**Comment by iakshay** [comment]
@mattgodbolt Can you summarize on items blocking this? I would really like to see Java support in compiler explorer. Thanks!

**Comment by mattgodbolt** [comment]
Closing in favour of #1843
